components/brave_wallet/browser/bitcoin/bitcoin_transaction_database.cc

```diff
@@ -0,0 +1,124 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_wallet/browser/bitcoin/bitcoin_transaction_database.h"
+
+#include <utility>
+
+#include "base/notreached.h"
+
+namespace brave_wallet {
+
+BitcoinTransactionDatabase::BitcoinTransactionDatabase() = default;
+BitcoinTransactionDatabase::~BitcoinTransactionDatabase() = default;
+
+void BitcoinTransactionDatabase::SetChainHeight(uint32_t chain_height) {
+  chain_height_ = chain_height;
+}
+
+absl::optional<uint32_t> BitcoinTransactionDatabase::GetChainHeight() const {
+  return chain_height_;
+}
+
+bool BitcoinTransactionDatabase::AddTransactions(
+    const std::string& address,
+    std::vector<bitcoin::Transaction> transactions) {
+  for (auto& tr : transactions) {
+    auto txid = tr.txid;
+
+    // TODO(apaymyshev): should check that transaction with same txid in set
+    // fully matches incoming one?
+
+    transactions_[address].insert(tr);
+  }
+
+  return true;
+}
+
+std::vector<bitcoin::Output> BitcoinTransactionDatabase::GetUnspentOutputs(
+    const std::string& address) {
+  std::map<bitcoin::Outpoint, const bitcoin::Output*> outputs;
+
+  for (auto& item : transactions_[address]) {
+    const bitcoin::Transaction& tr = item;
+    for (auto& o : tr.vout) {
+      if (o.scriptpubkey_address != address) {
+        continue;
+      }
+
+      outputs[o.outpoint] = &o;
+    }
+  }
+
+  for (auto& item : transactions_[address]) {
+    const bitcoin::Transaction& tr = item;
+    for (auto& i : tr.vin) {
```

Comment:
why can't we put this loop after for loop for vout so we don't have to iterate `transactions_[address]` twice?

Reply:
On the first run we go through all transactions related to address and search for outputs for that address, on the second run we search for inputs and remove these outputs. The ones remaining in the map are unspent outputs.
Refactored that bit to remove copy-paste and added comments 
https://github.com/brave/brave-core/pull/17978/commits/72d48e56e6778c7480f03b8d5fd63175c33eadf5

Reply:
I meant can’t we
```
for (const auto& tx : transactions) {
  for (const auto& o : tx.vout) {
  ...
  }
  for (const auto& i : tx.vin) {
  ...
  }
}
```
since we don’t remove or change anything in out loop

Reply:
as discussion on Slack,
```
There is no ordering for transactions, each of them just relates to that address
so we can’t see the chain of spending at only one iteration
```